PAE-266 Add ADR for approach to authenticating Admin UI with ADD (#138)
DEFRA/epr-backend: docs/architecture/decisions/0009-admin-ui-approach-for-authenticating-with-azure-ad.md

# 9. Admin-UI: Approach for authenticating with Azure AD

Date: 2025-10-02

## Status

Accepted

> ## Context

Author: I'm reluctant to commit to writing "(federated auth) involves a significantly more complex flow" because
- I don't understand it well enough to **quantify** how much _more_ complex it is
- I don't understand how much of that complexity will be abstracted from us anyways (though HAPI plugins etc)

I think capturing the fact that federated auth
- has more external dependencies
- requires more requests to external teams to get it set up

is sufficient to support the claim that secrets based will be quicker to deliver (and thus justify the decision for picking this approach)

As per the [Admin UI ADR](./0005-admin-ui-adr), the Admin UI will leverage Azure Active Directory (AAD) to present a single-sign-on journey to **authenticate** users trying to access the Admin UI. The Admin UI itself must authenticate with AAD to support the SSO interactions.

> At the time of writing, there are two documented/supported approaches

Reviewer: ```suggestion
At the time of writing, there are two supported approaches in the CDP platform.
```

> [!NOTE]

> > The Admin UI runs on the CDP platform (which is an abstracation over AWS), whereas AAD runs in (central) Defra's Azure account.

Reviewer: ```suggestion
> The Admin UI runs on the CDP platform (which is an abstraction over AWS), whereas AAD runs in (central) Defra's Azure account.
```

1. Supply `client-id` and `client-secret` to Admin UI app
1. At runtime the app uses `client-id` and `client-secret` to authenticate interactions with AAD
1. Periodically acquire a new `client-secret` (it expires)

```mermaid
sequenceDiagram

Actor U as User Browser
participant FE as Frontend
participant AAD as Azure AD

U->>FE: GET /some-page
Note over FE: Read session data
alt not logged in
  Note over FE: build AAD login url from config
  FE->>U: 302
  U->>AAD: GET https://login.microsoftonline.com/{tenantId}/oauth2/v2.0/authorize<br/>?client_id={client-id}&...
  AAD->>U: <html>Login form</html>
  Note over U: enter user name + password
  U->>AAD: Submit login form
  AAD->>U: 302
  U->>FE: GET /auth/callback<br/>?code={one-time-usage-code}<br/>&...
  FE->>AAD: Request user token<br/>(supplies {client-id}, {client-secret}, {one-time-usage-code})
  AAD->>FE: {JWT}
  Note over FE: Store JWT in session data
  FE->>U: 302: GET /some-page
else logged in
Note over FE: Read JWT from session data
  FE->>U: <html><br/>hello {username}<br/></html>
end
```

### Approach 2 - federated authentication

1. Register the Admin UI app with AAD, receive a `client-id`

> 1. Register the Admin UI app with AWS Congnito (on CDP)

Reviewer: ```suggestion
1. Register the Admin UI app with AWS Cognito (on CDP)
```
